C/tasks/north/sending_process/sending_process.cpp

```diff
@@ -0,0 +1,441 @@
+/*
+ * FogLAMP process class
+ *
+ * Copyright (c) 2018 Dianomic Systems
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Massimiliano Pinto
+ */
+
+#include <sending.h>
+#include <condition_variable>
+#include <reading_set.h>
+#include <plugin_manager.h>
+#include <plugin_api.h>
+#include <plugin.h>
+
+/**
+ * The sending process is run according to a schedule in order to send reading data
+ * to the historian, e.g. the PI system.
+ * It’s role is to implement the rules as to what needs to be sent and when,
+ * extract the data from the storage subsystem and stream it to the north
+ * for sending to the external system.
+ * The sending process does not implement the protocol used to send the data,
+ * that is devolved to the translation plugin in order to allow for flexibility
+ * in the translation process.
+ */
+
+using namespace std;
+
+// Buffer max elements
+#define DATA_BUFFER_ELMS 10
+
+// Thread sleep (milliseconds) when no data from storage layer
+#define TASK_FETCH_SLEEP 1000
+
+// historian plugin to load
+#define PLUGIN_NAME "omf"
+
+#define READINGS_FETCH_MAX 1000
+#define MAX_EXECUTION_TIME 60
+
+// Sendinf process configuration
+static const string sendingProcessConfiguration = "\"plugin\" : { \"type\" : \"string\", \"value\" : \"omf\", "
+							"\"default\" : \"omf\", \"description\" : \"Python module name of the plugin to load\" }" ;
+// Mutex for m_buffer access
+mutex      readMutex;
+// Mutex for thread idle time
+mutex	waitMutex;
+
+condition_variable cond_var;
+
+// Load data from storage
+static void loadDataThread(SendingProcess *loadData);
+// Send data from historian
+static void sendDataThread(SendingProcess *sendData);
+
+int main(int argc, char** argv)
```

Comment:
A lot of work is done in main before creating an object context. Personally I would not do it that way. Have a very small main and the rest in a sending process object context. Better still us the FogLAMP Process class approach to do it all. That is what it s there for.